[Dashboard] Fix error handling in "Save to library" modal for Links panel

olapawlus opened the pull request:
Closes #231164

When saving a Links panel to the library from the Dashboard, errors during the save process are not properly caught or displayed to the user.

This PR fixes the error handling and ensures that an error toast is shown to the user.

Video Demonstration of the fix:


https://github.com/user-attachments/assets/a376975c-267b-401f-ad25-d493aef0c7e2

olapawlus committed:
[Dashboard] Fix error handling in Save to library modal for links

olapawlus committed:
reverting unnecessary changes

olapawlus committed:
reverting unnecessary changes

nreese reviewed (commented):
Thanks for opening this issue and a PR to resolve the issue. I think the current solution resolves the issue too high in the call stack, and leaves other consumers of `showSaveModal` vulnerable to the same problem. For example, if you try to save a map with no network connect, the save model hangs just like this issue.

So instead of solving the problem for one instance in src/platform/plugins/private/links/public/content_management/save_to_library.tsx, lets solve the problem for all users and make `showSaveModal` more robust by putting the try/catch around `onSave` in src/platform/plugins/shared/saved_objects/public/save_modal/show_saved_object_save_modal.tsx

```
const onSaveConfirmed: MinimalSaveModalProps['onSave'] = async (...args) => {
        try {
          const response = await onSave(...args);
          // close modal if we either hit an error or the saved object got an id
          if (Boolean(isSuccess(response) ? response.id : response.error)) {
            closeModal();
          }
          return response;
        } catch(error) {
          closeModal();
          return { error };
        }
      };
```

Then links only needs one change, in src/platform/plugins/private/links/public/editor/get_editor_flyout.tsx, where `runSaveToLibrary` is called. `onSaveToLibrary` should throw if `saveResult.error`

```
onSaveToLibrary={async (newLinks: ResolvedLink[], newLayout: LinksLayoutType) => {
        const newState = {
          ...initialState,
          links: newLinks,
          layout: newLayout,
        };
        if (initialState?.savedObjectId) {
          const { savedObjectId, ...updateState } = newState;
          await linksClient.update({
            id: initialState.savedObjectId,
            data: {
              ...updateState,
              links: serializeResolvedLinks(newLinks),
            },
            options: { references: [] },
          });
          onCompleteEdit?.(newState);
          closeFlyout();
        } else {
          const saveResult = await runSaveToLibrary(newState);
          if (saveResult?.error) throw saveResult.error;
          onCompleteEdit?.(saveResult);
          // If saveResult is undefined, the user cancelled the save as modal and we should not close the flyout
          if (saveResult) closeFlyout();
        }
      }}
```

Then if wanted, try/catch could be removed from src/platform/plugins/private/links/public/content_management/save_to_library.tsx

olapawlus committed:
after cr adding try catch blocks

olapawlus commented:
> Thanks for opening this issue and a PR to resolve the issue. I think the current solution resolves the issue too high in the call stack, and leaves other consumers of `showSaveModal` vulnerable to the same problem. For example, if you try to save a map with no network connect, the save model hangs just like this issue.
> 
> So instead of solving the problem for one instance in src/platform/plugins/private/links/public/content_management/save_to_library.tsx, lets solve the problem for all users and make `showSaveModal` more robust by putting the try/catch around `onSave` in src/platform/plugins/shared/saved_objects/public/save_modal/show_saved_object_save_modal.tsx
> 
> ```
> const onSaveConfirmed: MinimalSaveModalProps['onSave'] = async (...args) => {
>         try {
>           const response = await onSave(...args);
>           // close modal if we either hit an error or the saved object got an id
>           if (Boolean(isSuccess(response) ? response.id : response.error)) {
>             closeModal();
>           }
>           return response;
>         } catch(error) {
>           closeModal();
>           return { error };
>         }
>       };
> ```
> 
> Then links only needs one change, in src/platform/plugins/private/links/public/editor/get_editor_flyout.tsx, where `runSaveToLibrary` is called. `onSaveToLibrary` should throw if `saveResult.error`
> 
> ```
> onSaveToLibrary={async (newLinks: ResolvedLink[], newLayout: LinksLayoutType) => {
>         const newState = {
>           ...initialState,
>           links: newLinks,
>           layout: newLayout,
>         };
>         if (initialState?.savedObjectId) {
>           const { savedObjectId, ...updateState } = newState;
>           await linksClient.update({
>             id: initialState.savedObjectId,
>             data: {
>               ...updateState,
>               links: serializeResolvedLinks(newLinks),
>             },
>             options: { references: [] },
>           });
>           onCompleteEdit?.(newState);
>           closeFlyout();
>         } else {
>           const saveResult = await runSaveToLibrary(newState);
>           if (saveResult?.error) throw saveResult.error;
>           onCompleteEdit?.(saveResult);
>           // If saveResult is undefined, the user cancelled the save as modal and we should not close the flyout
>           if (saveResult) closeFlyout();
>         }
>       }}
> ```
> 
> Then if wanted, try/catch could be removed from src/platform/plugins/private/links/public/content_management/save_to_library.tsx



> Thanks for opening this issue and a PR to resolve the issue. I think the current solution resolves the issue too high in the call stack, and leaves other consumers of `showSaveModal` vulnerable to the same problem. For example, if you try to save a map with no network connect, the save model hangs just like this issue.
> 
> So instead of solving the problem for one instance in src/platform/plugins/private/links/public/content_management/save_to_library.tsx, lets solve the problem for all users and make `showSaveModal` more robust by putting the try/catch around `onSave` in src/platform/plugins/shared/saved_objects/public/save_modal/show_saved_object_save_modal.tsx
> 
> ```
> const onSaveConfirmed: MinimalSaveModalProps['onSave'] = async (...args) => {
>         try {
>           const response = await onSave(...args);
>           // close modal if we either hit an error or the saved object got an id
>           if (Boolean(isSuccess(response) ? response.id : response.error)) {
>             closeModal();
>           }
>           return response;
>         } catch(error) {
>           closeModal();
>           return { error };
>         }
>       };
> ```
> 
> Then links only needs one change, in src/platform/plugins/private/links/public/editor/get_editor_flyout.tsx, where `runSaveToLibrary` is called. `onSaveToLibrary` should throw if `saveResult.error`
> 
> ```
> onSaveToLibrary={async (newLinks: ResolvedLink[], newLayout: LinksLayoutType) => {
>         const newState = {
>           ...initialState,
>           links: newLinks,
>           layout: newLayout,
>         };
>         if (initialState?.savedObjectId) {
>           const { savedObjectId, ...updateState } = newState;
>           await linksClient.update({
>             id: initialState.savedObjectId,
>             data: {
>               ...updateState,
>               links: serializeResolvedLinks(newLinks),
>             },
>             options: { references: [] },
>           });
>           onCompleteEdit?.(newState);
>           closeFlyout();
>         } else {
>           const saveResult = await runSaveToLibrary(newState);
>           if (saveResult?.error) throw saveResult.error;
>           onCompleteEdit?.(saveResult);
>           // If saveResult is undefined, the user cancelled the save as modal and we should not close the flyout
>           if (saveResult) closeFlyout();
>         }
>       }}
> ```
> 
> Then if wanted, try/catch could be removed from src/platform/plugins/private/links/public/content_management/save_to_library.tsx

Thanks so much for the really helpful explanation of the pattern with returning { error } and then throwing it higher up in the code!
I also added the same kind of handling in the `checkForDuplicateTitle`, because before the toast wasn’t showing if the error happened during the checkForDuplicateTitle function used in `saveToLibrary`. The toast also didn’t show when I removed the try-catch from `saveToLibrary`. Now the toast shows correctly.

olapawlus committed:
after cr adding try catch blocks

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/328404)
* Commit: 92263e8fde483dde98458bc0bb5124978a99e9f5

### Failed CI Steps
* [Jest Tests #19](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899b4-0fae-469c-801a-e88a8e1bc352)
* [Jest Tests #19](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899db-c278-40a9-af9f-ee095c46f8dc)
* [FTR Configs #66](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899b4-0ff5-4867-a707-a2273bf0aa83)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899b4-0fae-469c-801a-e88a8e1bc352) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899b4-0fae-469c-801a-e88a8e1bc352/artifacts/019899db-a547-4063-beb8-d45237e895fc) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should call batchBackfillRuleGaps correctly
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899db-c278-40a9-af9f-ee095c46f8dc) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899db-c278-40a9-af9f-ee095c46f8dc/artifacts/019899f2-525f-485b-8dfb-16c74f1ff478) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should call batchBackfillRuleGaps correctly
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899b4-0fae-469c-801a-e88a8e1bc352) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899b4-0fae-469c-801a-e88a8e1bc352/artifacts/019899db-a547-40ed-b257-57e52a592c40) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should ensure the user is authorized for each combination of ruleTypeId and consumer found in the rules
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899db-c278-40a9-af9f-ee095c46f8dc) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899db-c278-40a9-af9f-ee095c46f8dc/artifacts/019899f2-5260-48c5-80bd-f4f9962583f8) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should ensure the user is authorized for each combination of ruleTypeId and consumer found in the rules
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899b4-0fae-469c-801a-e88a8e1bc352) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899b4-0fae-469c-801a-e88a8e1bc352/artifacts/019899db-a547-405f-a789-fbb59cfd48d6) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should log an audit event when the rule fails at the authorization step
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899db-c278-40a9-af9f-ee095c46f8dc) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899db-c278-40a9-af9f-ee095c46f8dc/artifacts/019899f2-5260-4e81-8c65-769e6ee58fcc) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should log an audit event when the rule fails at the authorization step
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899b4-0fae-469c-801a-e88a8e1bc352) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899b4-0fae-469c-801a-e88a8e1bc352/artifacts/019899db-a546-44d7-8414-3c7fc6350700) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should refresh the index of the event client once after scheduling the backfilling of all gaps
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899db-c278-40a9-af9f-ee095c46f8dc) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899db-c278-40a9-af9f-ee095c46f8dc/artifacts/019899f2-5260-494b-8a5f-31a1ad50bcb9) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should refresh the index of the event client once after scheduling the backfilling of all gaps
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899b4-0fae-469c-801a-e88a8e1bc352) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899b4-0fae-469c-801a-e88a8e1bc352/artifacts/019899db-a548-42cc-b983-bb3f69825c52) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should return a list with errored rules
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899db-c278-40a9-af9f-ee095c46f8dc) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899db-c278-40a9-af9f-ee095c46f8dc/artifacts/019899f2-5261-41fe-af46-bb5c4c963f87) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should return a list with errored rules
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899b4-0fae-469c-801a-e88a8e1bc352) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899b4-0fae-469c-801a-e88a8e1bc352/artifacts/019899db-a545-4bc0-9663-a0a1310a126f) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should return a list with rules that succeeded
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899db-c278-40a9-af9f-ee095c46f8dc) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899db-c278-40a9-af9f-ee095c46f8dc/artifacts/019899f2-5260-4261-960f-8e21f0198523) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should return a list with rules that succeeded
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899b4-0fae-469c-801a-e88a8e1bc352) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899b4-0fae-469c-801a-e88a8e1bc352/artifacts/019899db-a547-48e7-9d35-ec0555bfe784) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should return a list with rules that were skipped
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899db-c278-40a9-af9f-ee095c46f8dc) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899db-c278-40a9-af9f-ee095c46f8dc/artifacts/019899f2-525f-4ec4-b0b3-4a4f2e3df162) Jest Tests #<span></span>19 / bulkFillGapsByRuleIds should return a list with rules that were skipped
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/328404#019899b4-0ff5-4867-a707-a2273bf0aa83) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/328404/jobs/019899b4-0ff5-4867-a707-a2273bf0aa83/artifacts/019899e3-2311-4d4b-bc70-0ed110fcb009) FTR Configs #<span></span>66 / discover/group6 discover unsaved changes badge should not show a badge after loading a saved search, only after changes

### History
* :broken_heart: [Build #328110](https://buildkite.com/elastic/kibana-pull-request/builds/328110) failed 532d4bd587531d345573571a5e10ac9c5bf7b943

cc @olapawlus

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/328404","number":328404,"commit":"92263e8fde483dde98458bc0bb5124978a99e9f5"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/328110","number":328110,"commit":"532d4bd587531d345573571a5e10ac9c5bf7b943"}],"number":328404}
buildkite-pr-comment-->

nreese reviewed (approved):
kibana-presentation changes LGTM
code review and tested in chrome

olapawlus committed:
Merge branch 'main' into errorHandlingLinks

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add a `backport:*` label or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 231168` locally
cc: @olapawlus
